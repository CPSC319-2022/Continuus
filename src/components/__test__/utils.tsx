import React, { type PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { configureStore, type PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";

import type { AppStore, RootState } from "~/redux/store";
import { reducer } from "~/redux/slices";
// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: reducer, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  // eslint-disable-next-line @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
