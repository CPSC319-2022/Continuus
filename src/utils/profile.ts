/**
 * Returns a path to the profile page for the given profile. If the argument is undefined, redirect to the home page.
 *
 * @param profile the user id of the account
 *
 * @return the profile path if the profile is a string, otherwise the home path
 */
export const userPathToProfile = (profile: string | undefined): string => {
    if (profile === undefined) {
        return '/';
    }

    return '/profile/' + encodeURIComponent(profile);
}
