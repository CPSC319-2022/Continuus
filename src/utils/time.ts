export const timeAgo = (inputDate: Date): string => {
  const now = new Date();
  const diffInSeconds = Math.max(Math.floor(
    (now.getTime() - inputDate.getTime()) / 1000
  ), 0);

  if (diffInSeconds < 60) {
    return `${diffInSeconds} ${diffInSeconds === 1 ? "second" : "seconds"} ago`;
  }

  const diffInMinutes = Math.floor(diffInSeconds / 60);
  if (diffInMinutes < 60) {
    return `${diffInMinutes} ${diffInMinutes === 1 ? "minute" : "minutes"} ago`;
  }

  const diffInHours = Math.floor(diffInMinutes / 60);
  if (diffInHours < 24) {
    return `${diffInHours} ${diffInHours === 1 ? "hour" : "hours"} ago`;
  }

  const diffInDays = Math.floor(diffInHours / 24);
  if (diffInDays < 30) {
    return `${diffInDays} ${diffInDays === 1 ? "day" : "days"} ago`;
  }

  const diffInMonths = Math.floor(diffInDays / 30);
  if (diffInMonths < 12) {
    return `${diffInMonths} ${diffInMonths === 1 ? 'month' : 'months'} ago`;
  }

  const diffInYears = Math.floor(diffInMonths / 12);
  return `${diffInYears} ${diffInYears === 1 ? 'year' : 'years'} ago`;
};
