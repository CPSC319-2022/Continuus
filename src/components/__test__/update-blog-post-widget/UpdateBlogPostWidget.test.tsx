import { describe, it, expect, beforeEach, vi } from "vitest";
import { cleanup, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithProviders } from "../utils";

const setup = (isSuccess: boolean) => {
  vi.mock("next-auth/react", () => ({
    useSession: () => ({
      data: {
        expires: "1",
        user: { id: "test-user-id", email: "a", name: "hi", image: "c" },
      },
      status: "authenticated",
    }),
  }));
  vi.doMock("~/utils/api", () => ({
    api: {
      blogPost: {
        update: {
          useMutation: () => ({
            onSuccess: vi.fn(),
            isSuccess,
          }),
        },
      },
      user: {
        currentUser: {
          useQuery: () => ({
            data: null,
            isLoading: false,
          }),
        },
      },
      useContext: () => ({
        blogPost: {
          get: {
            invalidate: vi.fn(),
          },
          count: {
            invalidate: vi.fn(),
          },
        },
      }),
    },
  }));
};

const props = {
  id: "test-id",
  title: "Test",
  content: "test",
};

describe("Update Blog Post Widget: Snapshot", () => {
  beforeEach(() => {
    cleanup();
    vi.clearAllMocks();
    vi.resetModules();
    vi.useRealTimers();
  });

  it("render widget", async () => {
    setup(false);
    const { UpdateBlogPostWidget } = await import(
      "~/components/update-blog-post-widget/UpdateBlogPostWidget"
    );
    const body = renderWithProviders(
      <UpdateBlogPostWidget {...props} />
    ).baseElement;
    expect(body).toMatchSnapshot();
  });

  it("update not successful", async () => {
    setup(false);
    const { UpdateBlogPostWidget } = await import(
      "~/components/update-blog-post-widget/UpdateBlogPostWidget"
    );
    const body = renderWithProviders(<UpdateBlogPostWidget {...props} />);
    await userEvent.click(screen.getByTestId("update-blog-post-button"));
    expect(body.baseElement).toMatchSnapshot();
  });

  it("update successful", async () => {
    setup(true);
    const { UpdateBlogPostWidget } = await import(
      "~/components/update-blog-post-widget/UpdateBlogPostWidget"
    );
    const body = renderWithProviders(<UpdateBlogPostWidget {...props} />);
    await userEvent.click(screen.getByTestId("update-blog-post-button"));
    
    await new Promise ((resolve) => setTimeout(() => resolve(true), 500));
    expect(body.baseElement).toMatchSnapshot();
  });
});
