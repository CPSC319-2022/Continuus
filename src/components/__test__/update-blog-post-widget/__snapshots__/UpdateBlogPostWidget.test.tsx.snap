// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Update Blog Post Widget: Snapshot > render widget 1`] = `
<body>
  <div>
    <div
      data-testid="update-blog-post-button"
    >
      <span>
        Edit
      </span>
    </div>
  </div>
  <div
    class="ReactModalPortal"
  />
</body>
`;

exports[`Update Blog Post Widget: Snapshot > render widget click button 1`] = `
<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div>
    <div
      data-testid="update-blog-post-button"
    >
      <span>
        Edit
      </span>
    </div>
  </div>
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay fixed inset-0 z-20 bg-black/75"
    >
      <div
        aria-label="Blog Post Input"
        aria-modal="true"
        class="ReactModal__Content absolute top-1/2 left-1/2 z-40 w-11/12 -translate-x-1/2 -translate-y-1/2 rounded-md bg-white md:w-1/2 p-4"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="flex pb-3 w-full flex-row items-center justify-between border-b border-b-gray-200 pl-4 pr-4"
        >
          <div
            class="font-semibold"
          >
            Update
             Blog Post
          </div>
        </div>
        <div
          class="flex h-96 w-full flex-col items-center justify-center"
        >
          <svg
            class="animate-pulse stroke-highlight-green"
            fill="none"
            height="12rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="12rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="20 6 9 17 4 12"
            />
          </svg>
          <div
            class="text-lg"
          >
            Successfully updated the blog post
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
</body>
`;
